File: test_xcloud_vs/python_driver/run_select_in_loop.py
#!/usr/bin/env python3
import sys
import time
from datetime import datetime
from cassandra.auth import PlainTextAuthProvider
from cassandra.cluster import Cluster

KEYSPACE_NAME = "myapp"
TABLE_NAME = "comments"

def log_print(message, log_file):
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    formatted = f"[{timestamp}] {message}"
    print(formatted)
    if log_file:
        print(formatted, file=log_file)
        log_file.flush()

def main():
    if len(sys.argv) != 4:
        print("Usage: python select_monitor.py <contact_point> <username> <password>")
        sys.exit(1)

    contact_point, username, password = sys.argv[1:4]
    log_filename = f"select_monitor_{datetime.now().strftime('%Y%m%d_%H%M%S')}.log"
    log_file = open(log_filename, 'w')

    log_print("Starting SELECT monitor script", log_file)
    log_print(f"Log file: {log_filename}", log_file)

    cluster = Cluster(
        contact_points=[contact_point],
        port=9042,
        auth_provider=PlainTextAuthProvider(username=username, password=password)
    )
    session = cluster.connect(KEYSPACE_NAME)

    while True:
        try:
            result = session.execute(f"SELECT COUNT(*) FROM {TABLE_NAME}")
            count = result.one()[0]
            log_print(f"SUCCESS: Row count in {TABLE_NAME}: {count}", log_file)
        except Exception as e:
            log_print(f"ERROR: {e}", log_file)
        time.sleep(5)

if __name__ == "__main__":
    main()

File: test_xcloud_vs/rust_driver/src/run_select_in_loop.rs
use scylla::{Session, SessionBuilder};
use std::env;
use std::fs::OpenOptions;
use std::io::Write;
use std::time::Duration;
use chrono::Local;

const KEYSPACE_NAME: &str = "myapp";
const TABLE_NAME: &str = "comments";

fn log_print(message: &str, log_file: &mut std::fs::File) {
    let timestamp = Local::now().format("%Y-%m-%d %H:%M:%S");
    let formatted = format!("[{}] {}", timestamp, message);
    println!("{}", formatted);
    writeln!(log_file, "{}", formatted).unwrap();
    log_file.flush().unwrap();
}

#[tokio::main]
async fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 4 {
        eprintln!("Usage: cargo run -- <contact_point> <username> <password>");
        std::process::exit(1);
    }
    let contact_point = &args[1];
    let username = &args[2];
    let password = &args[3];

    let log_filename = format!(
        "select_monitor_{}.log",
        Local::now().format("%Y%m%d_%H%M%S")
    );
    let mut log_file = OpenOptions::new()
        .create(true)
        .write(true)
        .append(true)
        .open(&log_filename)
        .expect("Cannot open log file");

    log_print("Starting SELECT monitor script", &mut log_file);
    log_print(&format!("Log file: {}", log_filename), &mut log_file);

    let session: Session = SessionBuilder::new()
        .known_node(contact_point)
        .user(username, password)
        .build()
        .await
        .expect("Failed to connect to ScyllaDB");

    session
        .use_keyspace(KEYSPACE_NAME, false)
        .await
        .expect("Failed to set keyspace");

    loop {
        match session
            .query(format!("SELECT COUNT(*) FROM {}", TABLE_NAME), &[])
            .await
        {
            Ok(res) => {
                let count: i64 = res.rows
                    .as_ref()
                    .and_then(|rows| rows.get(0))
                    .and_then(|row| row.columns[0].as_ref())
                    .and_then(|val| val.as_bigint())
                    .unwrap_or(0);
                log_print(
                    &format!("SUCCESS: Row count in {}: {}", TABLE_NAME, count),
                    &mut log_file,
                );
            }
            Err(e) => {
                log_print(&format!("ERROR: {}", e), &mut log_file);
            }
        }
        tokio::time::sleep(Duration::from_secs(5)).await;
    }
}
